users/users.go

```diff
@@ -1,228 +1,440 @@
+            return &rank
+        }
+    }
+    return nil
+}
+
+func GetNextRank(current *Rank) *Rank {
+    for i, rank := range Ranks {
+        if rank.Name == current.Name && i+1 < len(Ranks) {
+            return &Ranks[i+1]
+        }
+    }
+    return nil
+}
+
+func (user *User) UpdateRankIfNeeded() error {
+    // Load user events (Ban, Rejoin, etc.)
+    if err := user.LoadEvents(); err != nil {
+        log.Warnf("Could not load events for user %s: %v", user.GetName(), err)
+        return err
+    }
+
+    // Make sure user has a FirstWorkout recorded
+    if err := user.EnsureRankUpdatedAtExists(); err != nil {
+        return err
+    }
+
+    log.Infof("Start Rank Calculation for User %s (ID: %d)", user.GetName(), user.ID)
+
+    // Identify if the user had a Ban followed by a Rejoin
+    var lastBan *Event
+    var lastRejoin *Event
+
+    for i := len(user.Events) - 1; i >= 0; i-- {
+        event := user.Events[i]
+        if event.Event == BanEventType && lastBan == nil {
+            lastBan = &event
+        }
+        if event.Event == RejoinedGroupEventType && lastRejoin == nil {
+            lastRejoin = &event
+        }
+        if lastBan != nil && lastRejoin != nil {
+            break
+        }
+    }
+
+    // If there is a Rejoin after a Ban, reset RankUpdatedAt to the Rejoin date
+    if lastBan != nil && lastRejoin != nil && lastRejoin.CreatedAt.After(lastBan.CreatedAt) {
+        log.Infof("User %s has rejoined after ban. Resetting RankUpdatedAt to rejoin date: %s", user.GetName(), lastRejoin.CreatedAt.Format("2006-01-02"))
+        user.RankUpdatedAt = &lastRejoin.CreatedAt
+        if err := db.DBCon.Save(&user).Error; err != nil {
+            return err
+        }
+    }
+
+    // Ensure RankUpdatedAt exists
+    if user.RankUpdatedAt == nil {
+        log.Warnf("User %s has no RankUpdatedAt – cannot proceed with rank calculation.", user.GetName())
+        return nil
+    }
+
+    // Find the current rank
+    currentRank := GetRankByName(user.RankName)
+    if currentRank == nil {
+        log.Warnf("Unknown current rank '%s' for user %s. Defaulting to first rank.", user.RankName, user.GetName())
+        currentRank = &Ranks[0]
+    }
+
+    // Calculate effective days
+    var effectiveDays int
+
+    if lastBan != nil && (lastRejoin == nil || lastRejoin.CreatedAt.Before(lastBan.CreatedAt)) {
+        // User is banned and has not rejoined yet -> only MinDays are counted
+        log.Infof("User %s is currently banned. Using MinDays for effective days: %d", user.GetName(), currentRank.MinDays)
+        effectiveDays = currentRank.MinDays
+    } else {
+        // Normal case: user is active -> calculate days since RankUpdatedAt
+        effectiveDays = int(time.Since(*user.RankUpdatedAt).Hours() / 24)
+    }
+
+    log.Infof("Effective days for user %s: %d", user.GetName(), effectiveDays)
+
+    promoted := false
+
+    // Promote the user through ranks if enough days have passed
+    for {
```

Comment:
risk of infinite for loop with complex logic inside that can yield an error
let's add a limit of iterations?